web/api/query.go

```diff
@@ -48,6 +68,12 @@ func (serv MetricsService) Query(w http.ResponseWriter, r *http.Request) {
 	expr := params.Get("expr")
 	asText := params.Get("asText")
 
+	timestamp, err := parseTimestampOrNow(params.Get("timestamp"))
+	if err != nil {
+		http.Error(w, fmt.Sprintf("invalid query timestamp %s", err), http.StatusBadRequest)
```

Comment:
While you haven't set the content-type to application/json at this point already, that's very likely what the client expects. It should be a lot nicer for clients to return errors JSON encoded if that's the the media-type specified in Accept. Probably worth to add a error formatting function to be used in all such places which checks the Accept header and returns the error in the requested type.